javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -80,81 +80,283 @@ object Arrays {
   private def sortRangeImpl[@specialized T: ClassTag](
       a: Array[T], fromIndex: Int, toIndex: Int)(implicit ord: Ordering[T]): Unit = {
     checkIndicesForCopyOfRange(a.length, fromIndex, toIndex)
-    stableSortRangeImpl[T](a, fromIndex, toIndex - 1, new Array[T](a.length),
-        (a: T, b: T) => ord.lt(a, b))
+    quickSort[T](a, fromIndex, toIndex)
   }
 
   @inline
   private def sortRangeAnyRefImpl(a: Array[AnyRef], fromIndex: Int, toIndex: Int)(
       implicit ord: Ordering[AnyRef]): Unit = {
-    checkIndicesForCopyOfRange(a.length, fromIndex, toIndex)
-    stableSortRangeAnyRefImpl(a, fromIndex, toIndex - 1, new Array[AnyRef](a.length),
-      (a: AnyRef, b: AnyRef) => ord.lt(a, b))
+        checkIndicesForCopyOfRange(a.length, fromIndex, toIndex)
+    quickSortAnyRef(a, fromIndex, toIndex)
   }
 
   @inline
-  private def sortImpl[@specialized T: ClassTag](a: Array[T])(
-      implicit ord: Ordering[T]): Unit = {
-    stableSortRangeImpl[T](a, 0, a.length - 1, new Array[T](a.length), (a: T, b: T) => ord.lt(a, b))
-  }
+  private def sortImpl[@specialized T: ClassTag: Ordering](a: Array[T]): Unit =
+    quickSort[T](a, 0, a.length)
 
   @inline
-  private def sortAnyRefImpl(a: Array[AnyRef])(implicit ord: Ordering[AnyRef]): Unit = {
-    stableSortRangeAnyRefImpl(a, 0, a.length - 1, new Array[AnyRef](a.length), (a: AnyRef, b: AnyRef) => ord.lt(a, b))
+  private def sortAnyRefImpl(a: Array[AnyRef])(implicit ord: Ordering[AnyRef]): Unit =
+    quickSortAnyRef(a, 0, a.length)
+
+  // Implementation of sorting based on Scala 2.11.7 scala.util.Sorting
+  private final val qSortThreshold = 16
+
+  // Ordering[T] might be slow especially for boxed primitives, so use binary search variant of insertion sort
+  // Caller must pass iN >= i0 or math will fail.  Also, i0 >= 0.
+  @noinline
+  private final def insertionSort[@specialized T](a: Array[T], i0: Int, iN: Int)(implicit ord: Ordering[T]): Unit = {
```

Comment:
The `insertionSort` helper should be right after the `quickSort` it helps, not before. (Same for the `AnyRef` variant.)